### PR TITLE
Monitoring & Logging: Unify logging setup #4220

### PR DESCRIPTION
**Pull request by tbeerman**

### CODE REVIEW

**Comment by tbeerman** [comment]
@bari12 the release-patch-setup test will fail until you cherry-pick the flask dynamic endpoint loading patch into the release branch.

File: lib/rucio/common/logging.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+from rucio.common.config import config_get
+
+config_loglevel = getattr(logging, config_get('common', 'loglevel', raise_exception=False, default='DEBUG').upper())
+config_logformat = config_get('common', 'logformat', raise_exception=False, default='%(asctime)s\t%(name)s\t%(process)d\t%(levelname)s\t%(message)s')
+
+
+logging.basicConfig(stream=sys.stdout,
+                    level=config_loglevel,
+                    format=config_logformat)
```
Comment (bziemons) [inline lib/rucio/common/logging.py]: should be moved into a function and called from a function. To get rid of running code on import

File: etc/rucio.cfg.template [hunk 2]
```diff
@@ -1,6 +1,7 @@
 [common]
 logdir = /var/log/rucio
 loglevel = DEBUG
+logformat = %%(asctime)s\t%%(process)d\t%%(levelname)s\t%%(message)s
```
Comment (bziemons) [inline etc/rucio.cfg.template:4]: Does the `config_get` actually unquote the `%%`?

File: lib/rucio/core/nongrid_trace.py [hunk 3]
```diff
@@ -31,18 +31,34 @@
 from rucio.common.config import config_get, config_get_int
 from rucio.core.monitor import record_counter
 
-ERRLOG = logging.getLogger('errlog')
-ERRLOG.setLevel(logging.ERROR)
+CONFIG_COMMON_LOGLEVEL = getattr(logging, config_get('common', 'loglevel', raise_exception=False, default='DEBUG').upper())
+CONFIG_COMMON_LOGFORMAT = config_get('common', 'logformat', raise_exception=False, default='%(asctime)s\t%(process)d\t%(levelname)s\t%(message)s')
```
Comment (bziemons) [inline lib/rucio/core/nongrid_trace.py:35]: why do we need this here again? Can't we re-use this from `lib/rucio/common/logging.py`?

File: lib/rucio/core/trace.py [hunk 4]
```diff
@@ -37,18 +37,35 @@
 from rucio.common.config import config_get, config_get_int
 from rucio.core.monitor import record_counter
 
-ERRLOG = logging.getLogger('errlog')
-ERRLOG.setLevel(logging.ERROR)
+
+CONFIG_COMMON_LOGLEVEL = getattr(logging, config_get('common', 'loglevel', raise_exception=False, default='DEBUG').upper())
+CONFIG_COMMON_LOGFORMAT = config_get('common', 'logformat', raise_exception=False, default='%(asctime)s\t%(process)d\t%(levelname)s\t%(message)s')
```
Comment (bziemons) [inline lib/rucio/core/trace.py:42]: see above.

File: lib/rucio/daemons/conveyor/poller.py [hunk 5]
```diff
@@ -311,9 +306,9 @@ def poll_transfers(external_host, xfers, prepend_str='', request_ids=None, timeo
                     transfer_core.touch_transfer(external_host, transfer_id)
                 except (DatabaseException, DatabaseError) as error:
                     if re.match('.*ORA-00054.*', error.args[0]) or re.match('.*ORA-00060.*', error.args[0]) or 'ERROR 1205 (HY000)' in error.args[0]:
-                        logging.warning(prepend_str + "Lock detected when handling request %s - skipping" % request_id)
+                        logger(logging.WARNING, "Lock detected when handling request %s - skipping" % request_id)
                     else:
-                        logging.error(traceback.format_exc())
-            logging.debug(prepend_str + 'Finished updating %s transfer requests status (%i requests state changed) in %s seconds' % (len(xfers), cnt, (time.time() - tss)))
+                        logger(logging.ERROR, exc_info=True)
```
Comment (bziemons) [inline lib/rucio/daemons/conveyor/poller.py]: I think this needs a message, at least an empty string

File: lib/rucio/daemons/conveyor/poller.py [hunk 6]
```diff
@@ -311,9 +306,9 @@ def poll_transfers(external_host, xfers, prepend_str='', request_ids=None, timeo
...
+            logger(logging.DEBUG, 'Finished updating %s transfer requests status (%i requests state changed) in %s seconds' % (len(xfers), cnt, (time.time() - tss)))
     except Exception:
-        logging.error(traceback.format_exc())
+        logger(logging.ERROR, exc_info=True)
```
Comment (bziemons) [inline lib/rucio/daemons/conveyor/poller.py]: see above

File: lib/rucio/web/rest/flaskapi/v1/replicas.py [hunk 7]
```diff
@@ -178,7 +176,7 @@ def generate(vo):
         except RucioException as error:
             return generate_http_error_flask(500, error.__class__.__name__, error.args[0])
         except Exception as error:
-            print(format_exc())
+            logging.log(logging.ERROR, "Internal Error", exc_info=True)
```
Comment (bziemons) [inline lib/rucio/web/rest/flaskapi/v1/replicas.py]: why not use `logging.exception("Internal Error")` like everywhere?

File: lib/rucio/core/transfer.py [hunk 8]
```diff
@@ -1076,7 +1078,7 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
                     transfers[req_id]['use_ipv4'] = True
 
         except Exception:
-            logging.critical("Exception happened when trying to get transfer for request %s: %s" % (req_id, traceback.format_exc()))
+            logger(logging.CRITICAL, "Exception happened when trying to get transfer for request %s: %s" % (req_id, traceback.format_exc()))
```
Comment (bziemons) [inline lib/rucio/core/transfer.py]: use `exc_info=True` instead of `traceback.format_exc()`

File: lib/rucio/daemons/conveyor/common.py [hunk 9]
```diff
@@ -450,7 +446,7 @@ def bulk_group_transfer(transfers, policy='rule', group_bulk=200, source_strateg
     return grouped_jobs
 
 
-def get_conveyor_rses(rses=None, include_rses=None, exclude_rses=None, vos=None):
+def get_conveyor_rses(rses=None, include_rses=None, exclude_rses=None, vos=None, logger=logging.getLogger):
```
Comment (bziemons) [inline lib/rucio/daemons/conveyor/common.py]: this wouldn't work. This either needs some special handling or `logger=logging.log`.

idea for special handling (could also be done on a higher level, if useful):
```py
LOGGER = logging.getLogger(…)

def get_conveyor_rses(…, logger=LOGGER.log): 
```

**Review by bziemons — CHANGES_REQUESTED** [review]
Overall looks good. What bothers me is the logging setup plain in the logging module instead of a function. This is basically a replacement for the previous bad behavior on setting up the logger each time the logging module is imported.

**Comment by tbeerman** [comment]
The module only provides the formatted logger and the logging setup. So when you explicitly import the module you would anyway want logging and then the setup would be okay. I don't see the issue.

File: etc/rucio.cfg.template (hunk 2 above)
Comment (tbeerman) [reply etc/rucio.cfg.template:4]: Yes. With a single % it doesn't work. Another problem is the \t. It's not converted to an actual tab but it ends up as \t in the logs. I still have to figure out how to do that.

File: lib/rucio/core/nongrid_trace.py (hunk 3 above)
Comment (tbeerman) [reply lib/rucio/core/nongrid_trace.py:35]: The two trace modules are a bit special. I want a specific logger that write only the raw traces two a rotating log file as a buffer. The problem is, if I only setup the rotating logger, then these logs will also be duplicated to the root logger. That's why I added this hack to remove the handlers from the root logger and reconfigure it here, therefore I need these confit values again.

File: lib/rucio/daemons/conveyor/common.py (hunk 9 above)
Comment (tbeerman) [reply lib/rucio/daemons/conveyor/common.py]: Yeah, that's just a plain typo.

File: lib/rucio/web/rest/flaskapi/v1/replicas.py (hunk 7 above)
Comment (tbeerman) [reply lib/rucio/web/rest/flaskapi/v1/replicas.py]: Yeah, no idea why I did it differently here. Will fix it.

File: lib/rucio/tests/test_bin_rucio.py [hunk 10]
```diff
@@ -308,8 +308,8 @@ def test_upload_file_register_after_upload(self):
         cmd = 'rucio -v upload --rse {0} --scope {1} {2} {3} {4} --register-after-upload'.format(self.def_rse, self.user, tmp_file1, tmp_file2, tmp_file3)
         print(self.marker + cmd)
         exitcode, out, err = execute(cmd)
-        print(out)
-        print(err)
+        print("out")
+        print("err")
```
Comment (bziemons) [inline lib/rucio/tests/test_bin_rucio.py]: this doesn't make sense

**Review by bziemons — CHANGES_REQUESTED** [review]

**Review by bziemons — APPROVED** [review]

**Comment by bziemons** [comment]
Looks good, thanks!

Edit: also all tests against release-1.24 are running fine now as well.

**Comment by bari12** [comment]
This is extremely nice :-) Also thanks for testing @bziemons

**Pull request merged**